src/tunnel.cc

```diff
@@ -259,61 +266,90 @@ static CTCB tunnelTimeout;
 static EVH tunnelDelayedClientRead;
 static EVH tunnelDelayedServerRead;
 
+static bool
+doneWithTunnel(TunnelStateData *tunnelState)
+{
+    if (!tunnelState->noConnections())
+        return false;
+    // ConnStateData pipeline should contain the CONNECT we are performing
+    // but it may be invalid already (bug 4392)
+    if (tunnelState->http.valid() && tunnelState->http->getConn()) {
+        auto ctx = tunnelState->http->getConn()->pipeline.front();
+        if (ctx != nullptr)
+            ctx->finished();
+    }
+    delete tunnelState;
+    return true;
+}
+
 static void
-tunnelServerClosed(const CommCloseCbParams &params)
+doneWithServer(TunnelStateData *tunnelState)
 {
-    TunnelStateData *tunnelState = (TunnelStateData *)params.data;
-    debugs(26, 3, HERE << tunnelState->server.conn);
-    tunnelState->server.conn = NULL;
-    tunnelState->server.writer = NULL;
+    tunnelState->server.conn = nullptr;
+    tunnelState->server.writer = nullptr;
 
-    if (tunnelState->request != NULL)
+    if (tunnelState->request)
         tunnelState->request->hier.stopPeerClock(false);
-
-    if (tunnelState->noConnections()) {
-        // ConnStateData pipeline should contain the CONNECT we are performing
-        // but it may be invalid already (bug 4392)
-        if (tunnelState->http.valid() && tunnelState->http->getConn()) {
-            auto ctx = tunnelState->http->getConn()->pipeline.front();
-            if (ctx != nullptr)
-                ctx->finished();
-        }
-        delete tunnelState;
+    if (doneWithTunnel(tunnelState))
         return;
-    }
-
-    if (!tunnelState->client.writer) {
+    if (!tunnelState->client.writer)
         tunnelState->client.conn->close();
-        return;
-    }
 }
 
 static void
-tunnelClientClosed(const CommCloseCbParams &params)
+doneWithClient(TunnelStateData *tunnelState)
 {
-    TunnelStateData *tunnelState = (TunnelStateData *)params.data;
-    debugs(26, 3, HERE << tunnelState->client.conn);
-    tunnelState->client.conn = NULL;
-    tunnelState->client.writer = NULL;
+    tunnelState->client.conn = nullptr;
+    tunnelState->client.writer = nullptr;
 
-    if (tunnelState->noConnections()) {
-        // ConnStateData pipeline should contain the CONNECT we are performing
-        // but it may be invalid already (bug 4392)
-        if (tunnelState->http.valid() && tunnelState->http->getConn()) {
-            auto ctx = tunnelState->http->getConn()->pipeline.front();
-            if (ctx != nullptr)
-                ctx->finished();
-        }
-        delete tunnelState;
+    if (doneWithTunnel(tunnelState))
         return;
-    }
-
-    if (!tunnelState->server.writer) {
+    if (!tunnelState->server.writer)
         tunnelState->server.conn->close();
-        return;
+}
+
+/// Server close handler applied on the TLS connection being established.
+/// If TLS handshake failures, retries connection to another destination.
+static void
+serverClosed(const CommCloseCbParams &params)
```

Comment:
If this handler is specific to TLS closures, I would rename it to tlsServerClosed(), handleTlsServerClosure(), or similar.